### PR TITLE
Assignment #2

### PR DESCRIPTION
**Pull request by ustya-k**

### CODE REVIEW

File: week2/Q1.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+	def addRightChild(self,key):
+		self.right = BinaryTree(key)
+		
+	def addLeftChild(self,key):
+		self.left = BinaryTree(key)
+
+
+
+def get_ancestors(tree, key):
+	'''Returns ancestors of one key (not including itself).
```
Comment (flerdacodeu) [inline week2/Q1.py]: You can use the style of docstring in the example I shared for a template.You want to make sure you mention what it does, the arguments it takes, and the value it returns. Using "Args:" and "Results:" makes that easier to understand.

File: week2/Q2.py [hunk 2]
```diff
@@ -0,0 +1,65 @@
+class BinaryTree:
```
Comment (flerdacodeu) [inline week2/Q2.py]: This is the same class as above. Put it instead in a different module and import it into your two answers to avoid copying it. Code that is copied somewhere else needs to be updated multiple times...

File: week2/Q2.py [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+        self.root = key
+
+	def addRightChild(self,key):
+		self.right = BinaryTree(key)
+		
+	def addLeftChild(self,key):
+		self.left = BinaryTree(key)
+
+
+def get_ancestors(tree, key):
```
Comment (flerdacodeu) [inline week2/Q2.py]: This is also shared.

File: week2/Q2.py [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+
+def find_common_ancestor(ancestors1, ancestors2):
+    '''Finds lowest common ancestors given two lists of ancestors.
+    ancestors1 -- list of the ancestors of first given key (including the key)
+    ancestors2 -- list of the ancestors of second given key (including the key)
+    Returns None if at least one of the keys is not present in the tree.'''
+	counter = 0
+	len1 = len(ancestors1) - 1
+	len2 = len(ancestors2) - 1
+    lowest = None
```
Comment (flerdacodeu) [inline week2/Q2.py]: Indentation seems to be wrong (and indentation is a syntax error in Python).

File: week2/Q2.py [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+	while counter < len1 and counter < len2 and ancestors1[len1 - counter] == ancestors2[len2 - counter]:
```
Comment (flerdacodeu) [inline week2/Q2.py]: The use of the different counters is a bit confusing: how about having two counters `pos1` (and `pos2`) initialized at `len(ancestors1) - 1` and decremented each time? Basically you are iterating on the reversed list.

File: week2/Q2.py [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+		lowest = ancestors1[len1 - counter]
+		counter += 1
+	return lowest
```
Comment (flerdacodeu) [inline week2/Q2.py]: I believe there is a corner case (if one of the nodes is a parent of the other).
Let's add tests for this and other cases.

**Review by flerdacodeu — CHANGES_REQUESTED** [review]
Thanks Ustinia.

It would be helpful to write some tests for this code.

**Comment by ustya-k** [comment]
Thank you for the review! Sorry, for all the gaps, it's been tough two weeks.
I'm still struggling a bit with tests. I understand that first of all `return`s must be checked,  edge cases. But I do not know how to think about error cases. How to find good ones?

**Comment by flerdacodeu** [comment]
No problem at all Ustinia.

The point of this process is to help you learn how to do these things, no to check whether you are already able to do them, so we are good here.

I think you should start with something that represent what your function is meant to do: if your function returns the length of a list, check that, given a list with 3 elements, it returns 3. After that, you should check edge cases, like an empty list. Error cases may or may not apply to a specific function: for the list length function, you might want to check anything else -- the function assumes you pass it a list and there is nothing else you might want to pass or behavior you specify that makes sense; you could check that if you pass an int instead of a list, but really that should be obvious if, for example, the function is called `list_length(l)`; other functions define what happens in certain cases, for example, a function might want a positive integer or a non-empty list and say which exceptions are thrown in that case -- if that's the case, you can check for those cases too.

My suggestion is to start small and add more as you think of them. As your reviewer, I can also suggest something once there are some tests there.

Let me know if there is more info that would help you.

File: week2/Q1.py [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+
+	Args:
+		tree: BinaryTree, search is conducted on it 
+		key: element of a tree (type not specified), which ancestors algorithm is looking for
+	
+	Returns:
+		list, ancestors of the key, beginning from the closest
+		None, if the key is not present in the tree'''
+	if tree.root != key:
```
Comment (flerdacodeu) [inline week2/Q1.py]: Optional:
When you have an 'unbalanced' `if`-statement (especially with a return in one branch), it is usually more readable to have the shorter branch first.

Compare:
```python
  if [...]:
    [...]
    [...]
    [...]
    [...]
    [...]
  else:
    return [...]
```

with:

```python
  if not [...]:
    return [...]
  [...]
  [...]
  [...]
  [...]
  [...]
```

File: week2/Q2.py [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+		ancestors1: list, the ancestors of first given key (including the key)
+		ancestors2: list, the ancestors of second given key (including the key)
+
+	Returns:
+		element of a tree (type not specified)
+		None, if at least one of the keys is not present in the tree'''
+	pos1 = len(ancestors1) - 1
+	pos2 = len(ancestors2) - 1
+	lowest = None
+	while pos1 >= 0 and pos2 >= 0 and ancestors1[pos1] == ancestors2[pos2]:
```
Comment (flerdacodeu) [inline week2/Q2.py:18]: Optional: you can get done using a single counter that you update in the loop. Can you think of how?

File: week2/Q2.py [hunk 9]
```diff
@@ -0,0 +1,47 @@
+from btree import BinaryTree
+from Q1 import get_ancestors
+
+
+def find_common_ancestor(ancestors1, ancestors2):
```
Comment (flerdacodeu) [inline week2/Q2.py]: nit: This should probably be considered private. That is, it should be named `_find_common_ancestor`.

File: week2/btree.py [hunk 10]
```diff
@@ -0,0 +1,17 @@
+class BinaryTree:
+	'''Data structure, consists of root and two children which are trees as well (can be empty)
+		self.root: root element value
+		self.left: left child value
+		self.right: right child value
+		addRightChild: adds right child tree
```
Comment (flerdacodeu) [inline week2/btree.py]: nit: While fields are well documents as part of the class definition, methods are better documented in the definition of the method itself.

**Review by flerdacodeu — APPROVED** [review]
Thanks! Well done.
A few optional suggestions and consider.

File: week2/Q2.py (hunk 8 above)
Comment (ustya-k) [reply week2/Q2.py:18]: The way I was doing it before? Counter that counts number of elements from the end.
